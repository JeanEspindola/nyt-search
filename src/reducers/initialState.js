export default {
  list: [],
  page: 0,
  query: '',
};
