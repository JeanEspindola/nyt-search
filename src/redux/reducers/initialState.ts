/*
 * Initial State:
 *
 * It holds the initial state for the application store.
 * Page and query values are stored and passed back so the navigation buttons can use them
 * to make next or prev page call.
 *
 * */
import { InitialStateType, RootStateType } from './rootTypes'

export const initialState: InitialStateType = {
  list: [],
  page: 0,
  query: '',
  loading: false,
}

export const dummyRootState: RootStateType = {
  articleList: initialState,
}
