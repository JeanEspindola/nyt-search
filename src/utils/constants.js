export const API_KEY = 'de99f27226cb42b1993b11f717641745';
export const URL = 'https://api.nytimes.com/svc/search/v2/articlesearch.json';

export const SEARCH_SUBMIT_SUCCESS = 'SEARCH_SUBMIT_SUCCESS';
