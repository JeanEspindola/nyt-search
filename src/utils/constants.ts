export const API_KEY = 'aaassdfdfdfdfdfdf12343KApkAJmTE896X2'
export const URL = 'https://api.nytimes.com/svc/search/v2/articlesearch.json'
export const BASE_URL = `${URL}?api-key=${API_KEY}`

/* Filtering the search for only return articles */
export const FILTER_QUERY = 'document_type:("article")'
/* Restricts the response fields to only values necessary */
export const RESPONSE_FIELDS = 'snippet,pub_date,headline,web_url'
