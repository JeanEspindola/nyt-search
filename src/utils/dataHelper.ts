import { ArticleResponse } from '../redux/articleSearch/articleSearchTypes'

export const listAddKey = (articleList: ArticleResponse[]) => {
  return articleList.map(article => ({
    ...article,
    id: articleList.indexOf(article),
  }))
}
