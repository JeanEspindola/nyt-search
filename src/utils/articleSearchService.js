import 'whatwg-fetch';
import { URL, API_KEY, FILTER_QUERY, RESPONSE_FIELDS } from './constants';

const baseUrl = `${URL}?api-key=${API_KEY}`;

class ArticleSearchService {
  static getSearchResults(query, page) {
    const queryParams = `&q=${query}&page=${page}&fq=${FILTER_QUERY}&fl=${RESPONSE_FIELDS}`;
    const url = `${baseUrl}${queryParams}`;

    return fetch(url)
      .then(response => response.json())
      .then(json => json.response.docs)
      .catch(error => error);
  }
}

export default ArticleSearchService;
