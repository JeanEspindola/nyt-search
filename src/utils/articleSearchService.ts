/*
* Article Search Service:
*
* Functions available to use on the list. Separation of concerns.
*
* */
import 'whatwg-fetch';
import { URL, API_KEY, FILTER_QUERY, RESPONSE_FIELDS } from './constants';

const baseUrl = `${URL}?api-key=${API_KEY}`;

const ArticleSearchService = {
  /*
  * Here I'm using fetchAPI to make http requests.
  * */
  async getSearchResults(query: string, page: number) {
    const queryParams = `&q=${query}&page=${page}&fq=${FILTER_QUERY}&fl=${RESPONSE_FIELDS}`;
    const url = `${baseUrl}${queryParams}`;

    return fetch(url)
      .then(response => response.json())
      .then(json => json.response.docs)
      .catch(error => error);
  },

  /*
  * The returned data from the API does not have uniqueIDs. React demands one for rendering
  * lists, so before I return the data, I add id to each object in the array.
  * */
  listAddKey(array: Array<any>) {
    let id = 1;
    for (let it = 0; it < array.length; it += 1) {
      const item = array[it];
      item.id = id;
      id += 1;
    }

    return array;
  }
}

export default ArticleSearchService;
