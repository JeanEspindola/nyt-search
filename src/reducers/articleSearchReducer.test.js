import articleList from './articleSearchReducer';
import initialState from './initialState';
import { LOADING_INDICATOR_SUCCESS, SEARCH_SUBMIT_SUCCESS } from '../utils/constants';

describe('cartListReducer', () => {
  let action;

  it('Reducer --> SEARCH_SUBMIT_SUCCESS', () => {
    action = {
      list: [
        {
          article: 'a', snippet: 'a',
        },
        {
          article: 'b', snippet: 'b',
        },
        {
          article: 'c', snippet: 'c',
        },
      ],
      page: 3,
      query: 'test',
      type: SEARCH_SUBMIT_SUCCESS,
    };
    expect(articleList(initialState, action)).toMatchSnapshot();
  });

  it('Reducer --> LOADING_INDICATOR_SUCCESS', () => {
    action = {
      type: LOADING_INDICATOR_SUCCESS,
      loading: true,
    };

    expect(articleList(initialState, action)).toMatchSnapshot();
  });

  it('Reducer --> DEFAULT', () => {
    expect(articleList(initialState, { type: '' })).toMatchSnapshot();
  });

  it('Reducer --> return initial state', () => {
    expect(articleList(undefined, {})).toMatchSnapshot();
  });
});
