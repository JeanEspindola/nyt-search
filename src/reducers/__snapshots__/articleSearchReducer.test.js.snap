// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cartListReducer Reducer --> DEFAULT 1`] = `
Object {
  "list": Array [],
  "loading": false,
  "page": 0,
  "query": "",
}
`;

exports[`cartListReducer Reducer --> LOADING_INDICATOR_SUCCESS 1`] = `
Object {
  "list": Array [],
  "loading": true,
  "page": 0,
  "query": "",
}
`;

exports[`cartListReducer Reducer --> SEARCH_SUBMIT_SUCCESS 1`] = `
Object {
  "list": Array [
    Object {
      "article": "a",
      "id": 1,
      "snippet": "a",
    },
    Object {
      "article": "b",
      "id": 2,
      "snippet": "b",
    },
    Object {
      "article": "c",
      "id": 3,
      "snippet": "c",
    },
  ],
  "loading": false,
  "page": 3,
  "query": "test",
}
`;

exports[`cartListReducer Reducer --> return initial state 1`] = `
Object {
  "list": Array [],
  "loading": false,
  "page": 0,
  "query": "",
}
`;
